components/ee/payment-endpoint/src/chargebee/upgrade-helper.ts

```diff
@@ -42,4 +43,27 @@ export class UpgradeHelper {
             });
         });
     }
+
+    // Returns a ratio between 0 and 1:
+    //     0 means we've just finished the term
+    //     1 means we still have the entire term left
+    getCurrentTermRemainingRatio(chargebeeSubscription: chargebee.Subscription): number {
+        if (!chargebeeSubscription.next_billing_at) {
+            throw new Error('subscription.next_billing_at must be set.');
+        }
+        const now = new Date();
+        const nextBilling = new Date(chargebeeSubscription.next_billing_at * 1000);
+        const remainingMs = nextBilling.getTime() - now.getTime();
+
+        const unitToBillingPeriodInDays = { day: 1, week: 7, month: 365.25 / 12, year: 365.25 };
+        let billingPeriodInDays;
+        if (typeof chargebeeSubscription.billing_period === "number" && !!chargebeeSubscription.billing_period_unit) {
+            billingPeriodInDays = chargebeeSubscription.billing_period * unitToBillingPeriodInDays[chargebeeSubscription.billing_period_unit];
+        } else {
+            billingPeriodInDays = 1 * unitToBillingPeriodInDays["month"];
+        }
+        const billingPeriodMs = 1000 * 3600 * 24 * billingPeriodInDays;
+
+        return remainingMs / billingPeriodMs;
```

Comment:
FIXME: `remainingMs` (actual time left) can be longer than `billingPeriodMs` (average duration of a month) -- should make this more accurate, or prevent it from going over 1 month:

```
currentTermRemainingRatio=1.0184325058939843, diffInCents=2342"
```

(i.e. more than one month remaining, invoicing more than a full month of 23€)

Reply:
Fixed by getting the accurate total duration of the current billing cycle (some months are longer, some shorter, pro-rate should be based on actual cycle duration).